backend/backend/src/main/java/de/aaaaaaah/velcom/backend/newaccess/archiveaccess/ArchiveReadAccess.java

```diff
@@ -0,0 +1,165 @@
+package de.aaaaaaah.velcom.backend.newaccess.archiveaccess;
+
+import de.aaaaaaah.velcom.backend.newaccess.archiveaccess.exceptions.TarRetrieveException;
+import de.aaaaaaah.velcom.backend.newaccess.archiveaccess.exceptions.TarTransferException;
+import de.aaaaaaah.velcom.backend.newaccess.benchmarkaccess.entities.CommitSource;
+import de.aaaaaaah.velcom.backend.newaccess.committaccess.entities.CommitHash;
+import de.aaaaaaah.velcom.backend.newaccess.taskaccess.entities.Task;
+import de.aaaaaaah.velcom.backend.storage.repo.GuickCloning.CloneException;
+import de.aaaaaaah.velcom.backend.storage.repo.RepoStorage;
+import de.aaaaaaah.velcom.backend.storage.repo.exception.RepositoryAcquisitionException;
+import de.aaaaaaah.velcom.backend.storage.tar.TarFileStorage;
+import de.aaaaaaah.velcom.backend.util.TransferUtils;
+import de.aaaaaaah.velcom.shared.util.FileHelper;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.Random;
+import javax.annotation.Nullable;
+import org.eclipse.jgit.lib.AnyObjectId;
+import org.eclipse.jgit.lib.Constants;
+import org.eclipse.jgit.lib.Repository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An access for retrieving tar archives for tasks and for the bench repo.
+ */
+public class ArchiveReadAccess {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(ArchiveReadAccess.class);
+	private static final String BENCH_REPO_DIR_NAME = "benchrepo";
+
+	private final Path rootDir;
+	private final RepoStorage repoStorage;
+	private final TarFileStorage tarFileStorage;
+	private final String benchRepoUrl;
+
+	private final Random random;
+
+	public ArchiveReadAccess(Path rootDir, RepoStorage repoStorage, TarFileStorage tarFileStorage,
+		String benchRepoUrl) throws IOException {
+
+		this.rootDir = rootDir;
+		this.repoStorage = repoStorage;
+		this.tarFileStorage = tarFileStorage;
+		this.benchRepoUrl = benchRepoUrl;
+
+		random = new Random();
+
+		Files.createDirectories(rootDir);
+	}
+
+	/**
+	 * Transfer the tar file for a task to an {@link OutputStream}. Creates the tar file if
+	 * necessary.
+	 *
+	 * @param task the task whose tar file to transfer
+	 * @param outputStream the tar file is transferred to this stream
+	 * @throws TarRetrieveException if the tar file could not be retrieved
+	 * @throws TarTransferException if the tar file could not be transferred
+	 */
+	public void transferTask(Task task, OutputStream outputStream)
+		throws TarRetrieveException, TarTransferException {
+		if (task.getSource().getLeft().isPresent()) {
+			CommitSource commitSource = task.getSource().getLeft().get();
+			transferCommitTask(task, commitSource, outputStream);
+		} else {
+			transferTarTask(task, outputStream);
+		}
+	}
+
+	private Path getArchivePath(String repoDirName, CommitHash commitHash) {
```

Comment:
The thought behind it was that with multiple runners connected, there might be situations where more than one runner is served the same commit from the same repo, or maybe weird edge cases where the files from the old transfer have not yet been deleted before a new transfer starts. I'm relatively sure the random number is unnecessary though, so it might be best to just remove it.